modules/images/webp-uploads/load.php

```diff
@@ -284,11 +346,47 @@ function webp_uploads_remove_sources_files( $attachment_id ) {
 			}
 
 			$intermediate_file = str_replace( $basename, $properties['file'], $file );
-			if ( ! empty( $intermediate_file ) ) {
-				$intermediate_file = path_join( $upload_path['basedir'], $intermediate_file );
-				wp_delete_file_from_directory( $intermediate_file, $intermediate_dir );
+			if ( empty( $intermediate_file ) ) {
+				continue;
+			}
+
+			$intermediate_file = path_join( $upload_path['basedir'], $intermediate_file );
+			if ( ! file_exists( $intermediate_file ) ) {
+				continue;
 			}
+
+			wp_delete_file_from_directory( $intermediate_file, $intermediate_dir );
+		}
+	}
+
+	if ( ! isset( $metadata['sources'] ) || ! is_array( $metadata['sources'] ) ) {
+		return;
+	}
+
+	$original_mime_from_post = get_post_mime_type( $attachment_id );
+	$original_mime_from_file = wp_check_filetype( $file )['type'];
+
+	// Delete full sizes mime types.
+	foreach ( $metadata['sources'] as $mime => $properties ) {
+		// Don't remove the image with the same mime type as the original image as this would be removed by WordPress.
+		if ( $mime === $original_mime_from_post || $mime === $original_mime_from_file ) {
+			continue;
+		}
```

Comment:
This is overkill, checking the MIME type from the post will be sufficient. In the existing code for the sub-sizes we do the same thing, only checking the metadata, but not the file itself.
```suggestion
	$original_mime = get_post_mime_type( $attachment_id );

	// Delete full sizes mime types.
	foreach ( $metadata['sources'] as $mime => $properties ) {
		// Don't remove the image with the same mime type as the original image as this would be removed by WordPress.
		if ( $mime === $original_mime ) {
			continue;
		}
```

Reply:
That's a fair point, in the subsizes is determined during the upload however exists a use case when the mime type of the file and the `post->mime_type` won't be the same, specifically when using the filter that exists en core and the images are uploaded as WebP the `post->mime_type` would be set as JPEG when the actual mime type is WebP if that's the case duplicated WebP images would be generated.

Details can be found here: https://github.com/WordPress/performance/issues/75

This is not the same for the subsizes as at the time the mime property for the subsizes is set we can rely in the fact that is the correct one as it creates the mime after the file was uploaded as WebP. Agree this is an edge case but worth consider it as this an edge case to consider.

Reply:
Thanks for clarifying, that makes sense!